### PR TITLE
html: Add a test for form.requestSubmit()

### PR DESCRIPTION
**Pull request by tkent-google**
Specification issue: https://github.com/whatwg/html/issues/4187
Specification PR: https://github.com/whatwg/html/pull/4597

### CODE REVIEW

File: html/semantics/forms/the-form-element/form-requestsubmit.html [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    form.requestSubmit(submitButton);
+  });
+}, 'Passing a submit button not owned by the context object should throw');
+
+test(() => {
+  document.body.insertAdjacentHTML('afterbegin', `<form target="_blank">
+<button type="submit"></button>
+<input type="submit">
+<input type="image">
+<input type="reset">
```
Comment (zcorpan) [inline html/semantics/forms/the-form-element/form-requestsubmit.html]: This one is not used

File: html/semantics/forms/the-form-element/form-requestsubmit.html [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  let submitButton = document.createElement('button');
+  submitButton.type = 'submit';
+  assert_throws(new TypeError(), () => {
+    form.requestSubmit(submitButton);
+  });
+}, 'Passing a submit button not owned by the context object should throw');
+
+test(() => {
+  document.body.insertAdjacentHTML('afterbegin', `<form target="_blank">
+<button type="submit"></button>
```
Comment (zcorpan) [inline html/semantics/forms/the-form-element/form-requestsubmit.html]: Also test button without a type attribute and one with an invalid value

File: html/semantics/forms/the-form-element/form-requestsubmit.html [hunk 3]
```diff
@@ -0,0 +1,84 @@
+<!DOCTYPE html>
+<body>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+
+<iframe name="iframe" src="about:blank"></iframe>
+
+<script>
+test(() => {
+  document.body.insertAdjacentHTML('afterbegin', '<form><input type="reset"></form>');
```
Comment (zcorpan) [inline html/semantics/forms/the-form-element/form-requestsubmit.html]: Also test button element with type=button and type=reset

File: html/semantics/forms/the-form-element/form-requestsubmit.html [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+test(() => {
+  document.body.insertAdjacentHTML('afterbegin', '<form><input required></form>');
+  let form = document.querySelector('form');
+  let invalidControl = form.querySelector('input:invalid');
+  // Make sure no focused element.
+  document.activeElement.blur();
+  assert_not_equals(document.activeElement, invalidControl);
+
+  form.requestSubmit();
+  assert_equals(document.activeElement, invalidControl);
```
Comment (zcorpan) [inline html/semantics/forms/the-form-element/form-requestsubmit.html]: Focusing is only a "may" per spec. Check the `invalid` event instead?

https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#interactively-validate-the-constraints

File: html/semantics/forms/the-form-element/form-requestsubmit.html [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+}, 'requestSubmit() should trigger interactive form validation');
+
+window.onload = () => {
+  async_test('The value of the submitter should be appended, and form* ' +
```
Comment (zcorpan) [inline html/semantics/forms/the-form-element/form-requestsubmit.html]: Isn't the argument order the other way around for async_test?

File: html/semantics/forms/the-form-element/form-requestsubmit.html [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+test(() => {
+  document.body.insertAdjacentHTML('afterbegin', '<form><input type="reset"></form>');
+  let form = document.querySelector('form');
+  let submitButton = document.createElement('button');
+  submitButton.type = 'submit';
+  assert_throws(new TypeError(), () => {
+    form.requestSubmit(submitButton);
+  });
+}, 'Passing a submit button not owned by the context object should throw');
```
Comment (zcorpan) [inline html/semantics/forms/the-form-element/form-requestsubmit.html]: The element isn't even in the same tree, here. There are other interesting tests as well.

* The control is a descendant, but has a `form` attribute overriding the form owner to a different form.
* The control is a descendant, but has a `form` attribute that gives the control *no* form owner (e.g. `form=""`)
* The control is a descendant, but is actually owned by a different form by association from the parser. [1] or [2]

For each case, check that `A.requestSubmit(control)` throws, and `B.requestSubmit(control)` works.

[1] nested form actually owns the control
```
<!DOCTYPE html>
<form action=data:,A>
<div>
</form>
<form action=data:,B>
<input type=submit>
</form>
</div>
```
[2] the control is a descendant of the A form, not the B form, but is actually owned by B
```
<!DOCTYPE html>
<form action=data:,A>
<table>
</form>
<form action=data:,B>
<tr><td><input type=submit></td></tr>
</form>
</table>
```

File: html/semantics/forms/the-form-element/form-requestsubmit.html [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    let iframe = document.body.querySelector('iframe');
+    assert_true(form.matches(':invalid'), 'The form is invalid.');
+    // The form should be submitted though it is invalid.
+    iframe.addEventListener('load', t.step_func_done(() => {
+      assert_not_equals(iframe.src.indexOf('s=v'), -1);
+    }));
+    form.requestSubmit(form.querySelector('[type=submit]'));
+  });
+};
+</script>
```
Comment (zcorpan) [inline html/semantics/forms/the-form-element/form-requestsubmit.html]: Also test what happens when the entire form is not connected to the document

File: html/semantics/forms/the-form-element/form-requestsubmit.html [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+test(() => {
+  document.body.insertAdjacentHTML('afterbegin', '<form><input type="reset"></form>');
+  let form = document.querySelector('form');
+  assert_throws(new TypeError(), () => {
+    form.requestSubmit(document.body);
+  });
+  assert_throws(new TypeError(), () => {
+    form.requestSubmit(form.querySelector('input[type=reset]'));
+  });
+}, 'Passing an element which is not a submit button should throw');
```
Comment (zcorpan) [inline html/semantics/forms/the-form-element/form-requestsubmit.html]: A reset button is also not a submittable element. Also test a submittable element that is not a submit button (like `input type=text`)

File: html/semantics/forms/the-form-element/form-requestsubmit.html [hunk 9]
```diff
@@ -0,0 +1,84 @@
+<!DOCTYPE html>
+<body>
```
Comment (zcorpan) [inline html/semantics/forms/the-form-element/form-requestsubmit.html]: Missing `<title>`. Also, why add the `<body>` here (testharness.js usually goes in head)?

**Review by zcorpan — CHANGES_REQUESTED** [review]

**Comment by zcorpan** [comment]
https://github.com/whatwg/html/pull/4597#discussion_r282452692 also seems interesting to test

File: html/semantics/forms/the-form-element/form-requestsubmit.html (hunk 9 above)
Comment (tkent-google) [reply html/semantics/forms/the-form-element/form-requestsubmit.html]: Added <title>, removed <body>

File: html/semantics/forms/the-form-element/form-requestsubmit.html (hunk 3 above)
Comment (tkent-google) [reply html/semantics/forms/the-form-element/form-requestsubmit.html]: Done.

File: html/semantics/forms/the-form-element/form-requestsubmit.html (hunk 8 above)
Comment (tkent-google) [reply html/semantics/forms/the-form-element/form-requestsubmit.html]: Done.

File: html/semantics/forms/the-form-element/form-requestsubmit.html (hunk 6 above)
Comment (tkent-google) [reply html/semantics/forms/the-form-element/form-requestsubmit.html]: This is not a test of ```.form``` IDL attribute, and IMO  testing various patterns of ```.form``` here isn't cost-effective.  I added a few cases which ensures UAs don't apply simple ancestor check.

File: html/semantics/forms/the-form-element/form-requestsubmit.html (hunk 2 above)
Comment (tkent-google) [reply html/semantics/forms/the-form-element/form-requestsubmit.html]: Done.

File: html/semantics/forms/the-form-element/form-requestsubmit.html (hunk 1 above)
Comment (tkent-google) [reply html/semantics/forms/the-form-element/form-requestsubmit.html]: Removed.

File: html/semantics/forms/the-form-element/form-requestsubmit.html (hunk 4 above)
Comment (tkent-google) [reply html/semantics/forms/the-form-element/form-requestsubmit.html]: Good point.  I switched to ```invalid``` event.

File: html/semantics/forms/the-form-element/form-requestsubmit.html (hunk 5 above)
Comment (tkent-google) [reply html/semantics/forms/the-form-element/form-requestsubmit.html]: Wow, my bad.  Fixed.

File: html/semantics/forms/the-form-element/form-requestsubmit.html (hunk 7 above)
Comment (tkent-google) [reply html/semantics/forms/the-form-element/form-requestsubmit.html]: Done.

**Comment by tkent-google** [comment]
> whatwg/html#4597 (comment) also seems interesting to test

Added testcases based on https://github.com/whatwg/html/pull/4621 .

**Review by zcorpan — APPROVED** [review]

**Pull request merged**